### PR TITLE
Webhook configurations can choose which version of Review request they accept

### PR DESCRIPTION
**Pull request by mbohlool**
This is a [task from Extensibility GA](https://github.com/liggitt/enhancements/blob/7b7fc0c2293c02c9bb580805dbdae49d802a189b/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#admissionreview-v1)

Currently apiserver sends AdmissionReview to Admission Webhooks in v1beta1. We are going to introduce AdmissionReview v1 and a prerequisite for that is to have a mechanism for the webhook to define versions they understand. This PR addresses that.

The mechanism is also added to CRD Conversion Webhook stack.

```release-note
- Add mechanism for Admission Webhooks to specify which version of AdmissionReview they support
- Add mechanism for CRD Conversion Webhooks to specify which version of ConversionReview they support
```

/cc @liggitt @roycaihw @sttts @caesarxuchao

### CODE REVIEW

File: pkg/apis/admissionregistration/validation/validation.go [hunk 1]
```diff
@@ -150,10 +150,36 @@ func validateRule(rule *admissionregistration.Rule, fldPath *field.Path, allowSu
...
+func validateAdmissionReviewVersions(versions []string, fldPath *field.Path) field.ErrorList {
+	allErrors := field.ErrorList{}
+
+	// Currently only v1beta1 accepted in AdmissionReviewVersions
+	if len(versions) < 1 {
+		allErrors = append(allErrors, field.Required(fldPath, ""))
+	} else {
+		seen := map[string]bool{}
+		for i, v := range versions {
+			if !isAcceptedAdmissionReviewVersion(v) {
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: we don't want to prevent versions we don't understand, or that will block a webhook that supports v1beta1 and v1 from registering itself

File: pkg/apis/admissionregistration/validation/validation.go [hunk 2]
```diff
@@ -150,10 +150,36 @@ func validateRule(rule *admissionregistration.Rule, fldPath *field.Path, allowSu
 	return allErrors
 }
 
+func isAcceptedAdmissionReviewVersion(v string) bool {
+	return "v1beta1" == v
+}
+
+func validateAdmissionReviewVersions(versions []string, fldPath *field.Path) field.ErrorList {
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: this function is still getting called on update (`*Strategy#ValidateUpdate` -> `ValidateValidatingWebhookConfiguration` -> `validateAdmissionReviewVersions`), but lacks the information to know whether it should require a recognized version

we should limit `*Strategy#ValidateUpdate` to calling `Validate*Update`, and not have the strategy also call the `Validate*` method directly (that would let us unit test `Validate*` and `Validate*Update` methods and have that actually reflect what validation would be done on create and update)

File: pkg/apis/admissionregistration/validation/validation.go [hunk 3]
```diff
@@ -248,9 +275,47 @@ func validateRuleWithOperations(ruleWithOperations *admissionregistration.RuleWi
...
+	for _, hook := range oldC.Webhooks {
+		oldWebhooks[hook.Name] = hook
+	}
+
+	for i, hook := range newC.Webhooks {
+		fldPath := field.NewPath("webhooks").Index(i)
+		allErrors = append(allErrors, validateWebhook(&hook, fldPath)...)
+		oldHook, exists := oldWebhooks[hook.Name]
+		if exists {
+			// Only validate admissionReviewVersions if they were valid before
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: skipping this if the old webhook didn't have recognized versions means we also skip required and duplicate checking, which isn't great

Comment (liggitt) [reply pkg/apis/admissionregistration/validation/validation.go]: it might be simpler to do this at the object level rather than the webhook level (being more fine-grained doesn't do a lot for us, and makes the code way more complicated):

* set hadRecognizedVersions to true if all webhooks in the old object provided recognized versions
* change validateAdmissionReviewVersions to take `requireRecognizedVersion bool` and pass in hadRecognizedVersions here, and pass in `true` from `ValidateMutatingWebhookConfiguration` (you did something similar in validateCustomResourceDefinitionSpec and it worked well)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go [hunk 4]
```diff
@@ -66,6 +66,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 					Strategy: apiextensions.NoneConverter,
 				}
 			}
+			obj.Conversion.ConversionReviewVersions = []string{"v1beta1"}
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go]: only if len == 0?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go [hunk 5]
```diff
@@ -68,6 +68,9 @@ func SetDefaults_CustomResourceDefinitionSpec(obj *CustomResourceDefinitionSpec)
 			Strategy: NoneConverter,
 		}
 	}
+	if len(obj.Conversion.ConversionReviewVersions) == 0 {
+		obj.Conversion.ConversionReviewVersions = []string{"v1beta1"}
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go]: prefer SchemeGroupVersion.Version

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 6]
```diff
@@ -108,6 +108,11 @@ func (d *validatingDispatcher) callHook(ctx context.Context, h *v1beta1.Webhook,
 		}
 	}
 
+	// Currently dispatcher only supports `v1beta1` AdmissionReview
+	if !util.HasAdmissionReviewVersion("v1beta1", h) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go]: prefer v1beta1.SchemeGroupVersion.Version, and add a TODO to make the dispatcher capable of sending/receiving multiple versions

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 7]
```diff
@@ -94,6 +94,11 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
 		}
 	}
 
+	// Currently dispatcher only supports `v1beta1` AdmissionReview
+	if !util.HasAdmissionReviewVersion("v1beta1", h) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: use constant, and add TODO to make this capable of sending/receiving multiple versions

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go [hunk 8]
```diff
@@ -216,6 +230,11 @@ func (c *webhookConverter) ConvertToVersion(in runtime.Object, target runtime.Gr
 		}
 	}
 
+	// Currently converter only supports `v1beta1` ConversionReview
+	if !c.hasConversionReviewVersion("v1beta1") {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go]: use constant and add TODO

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 9]
```diff
@@ -250,6 +283,9 @@ func ValidateCustomResourceConversion(conversion *apiextensions.CustomResourceCo
 				allErrs = append(allErrs, webhook.ValidateWebhookService(fldPath.Child("webhookClientConfig").Child("service"), cc.Service.Name, cc.Service.Namespace, cc.Service.Path)...)
 			}
 		}
+		if mustValidateConversionReviewVersions {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: required and duplicate checking is still worthwhile in all cases, this should only govern whether we require a recognized version

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go (hunk 4 above)
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go]: Hmm, is the point of this function to make the object valid? any value other than []string{"v1beta1"} is not valid.

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go]: fuzzing doesn't care about validation, just about round-trippability. len(0) doesn't round-trip because empty lists get defaulted

File: pkg/apis/admissionregistration/validation/validation.go (hunk 2 above)
Comment (mbohlool) [reply pkg/apis/admissionregistration/validation/validation.go]: ValidateUpdate should not call Validate*. If you check current code, both methods it calls doing the same thing and we basically duplicate validation errors.

`strategy#ValidateUpdate` calls to both `ValidateValidatingWebhookConfiguration` and `ValidateValidatingWebhookConfigurationUpdate`. `ValidateValidatingWebhookConfigurationUpdate` calls `ValidateValidatingWebhookConfiguration` on the newObject which is the same as first call `strategy#ValidateUpdate` did to `ValidateValidatingWebhookConfiguration`. I will fix strategy.

File: pkg/apis/admissionregistration/validation/validation.go [hunk 10]
```diff
@@ -247,10 +283,25 @@ func validateRuleWithOperations(ruleWithOperations *admissionregistration.RuleWi
 	return allErrors
 }
 
+// hasUnrecognizedVersions returns true of any of the webhooks has an unrecognized version in AdmissionReviewVersions.
+func hasUnrecognizedAdmissionReviewVersions(webhooks []admissionregistration.Webhook) bool {
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: function name is misleading, since this can return false for webhooks which do have some unrecognized versions. `hasRecognizedAdmissionReviewVersions` would be clearer, returning true if the existing set of webhooks all had recognized versions.

File: pkg/apis/admissionregistration/validation/validation.go [hunk 11]
```diff
@@ -247,10 +283,25 @@ func validateRuleWithOperations(ruleWithOperations *admissionregistration.RuleWi
...
+	unusedFieldPath := field.NewPath("webhooks")
+	for _, hook := range webhooks {
+		// only skip checking for recognized version if we had invalid versions
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: calling a validation function which validates several things twice, with different params, then using different results to infer a particular condition is hard to follow.

would be simpler to just check for a supported version in each webhook directly:
```go
// hasAcceptedAdmissionReviewVersions returns true if all webhooks have at least one
// admission review version this apiserver accepts.
func hasAcceptedAdmissionReviewVersions(webhooks []admissionregistration.Webhook) bool {
  for _, hook := range webhooks {
    hasRecognizedVersion := false
    for _, version := range hook.AdmissionReviewVersions {
      if isAcceptedAdmissionReviewVersion(version) {
        hasRecognizedVersion = true
        break
      }
    }
    if !hasRecognizedVersion {
      return false
    }
  }
  return true
}
```

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 12]
```diff
@@ -258,7 +293,14 @@ func ValidateCustomResourceConversion(conversion *apiextensions.CustomResourceCo
 
 // ValidateCustomResourceDefinitionSpecUpdate statically validates
 func ValidateCustomResourceDefinitionSpecUpdate(spec, oldSpec *apiextensions.CustomResourceDefinitionSpec, established bool, fldPath *field.Path) field.ErrorList {
-	allErrs := ValidateCustomResourceDefinitionSpec(spec, fldPath)
+	requireRecognizedVersion := true
+	if oldSpec.Conversion != nil {
+		// only skip requiring recognized versions if we had an unrecognized version, not for any other invalid values.
+		wasValidExceptInvalidVersions := len(validateConversionReviewVersions(oldSpec.Conversion.ConversionReviewVersions, false, fldPath)) == 0
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: same comment about checking for an accepted version directly instead of calling validation twice and inferring from the different results

File: pkg/apis/admissionregistration/validation/validation.go [hunk 13]
```diff
@@ -150,18 +151,53 @@ func validateRule(rule *admissionregistration.Rule, fldPath *field.Path, allowSu
...
+func validateValidatingWebhookConfiguration(e *admissionregistration.ValidatingWebhookConfiguration, requireRecognizedVersion bool) field.ErrorList {
 	allErrors := genericvalidation.ValidateObjectMeta(&e.ObjectMeta, false, genericvalidation.NameIsDNSSubdomain, field.NewPath("metadata"))
 	for i, hook := range e.Webhooks {
 		allErrors = append(allErrors, validateWebhook(&hook, field.NewPath("webhooks").Index(i))...)
+		allErrors = append(allErrors, validateAdmissionReviewVersions(hook.AdmissionReviewVersions, requireRecognizedVersion, field.NewPath("webhooks").Index(i).Child("admissionReviewVersions"))...)
 	}
 	return allErrors
 }
 
 func ValidateMutatingWebhookConfiguration(e *admissionregistration.MutatingWebhookConfiguration) field.ErrorList {
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: mutatingWebhookConfigurationStrategy#ValidateUpdate still calls this, and should not

Comment (mbohlool) [reply pkg/apis/admissionregistration/validation/validation.go]: fixed

File: pkg/apis/admissionregistration/validation/validation.go (hunk 11 above)
Comment (mbohlool) [reply pkg/apis/admissionregistration/validation/validation.go]: Hmm, I just realized the validation may have been wrong anyway till now.  A webhook is valid if it has at least one recognized version, right? The validation right now requires all valid values. will fix.

**Comment by mbohlool** [comment]
@liggitt fixed validations. PTAL.

File: pkg/apis/admissionregistration/validation/validation.go [hunk 14]
```diff
@@ -150,18 +151,57 @@ func validateRule(rule *admissionregistration.Rule, fldPath *field.Path, allowSu
...
+func validateAdmissionReviewVersions(versions []string, requireRecognizedVersion bool, fldPath *field.Path) field.ErrorList {
+	allErrors := field.ErrorList{}
+
+	// Currently only v1beta1 accepted in AdmissionReviewVersions
+	if len(versions) < 1 {
+		allErrors = append(allErrors, field.Required(fldPath, ""))
+	} else {
+		seen := map[string]bool{}
+		hasAcceptedVersion := false
+		for i, v := range versions {
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: should probably constrain the version strings to `utilvalidation.IsDNS1035Label` (here and when validating conversionReviewVersions), since that's what we limit APIService versions to:
```
	for _, errString := range utilvalidation.IsDNS1035Label(apiService.Spec.Version) {
		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "version"), apiService.Spec.Version, errString))
	}
```

File: pkg/apis/admissionregistration/validation/validation.go [hunk 15]
```diff
@@ -150,18 +151,57 @@ func validateRule(rule *admissionregistration.Rule, fldPath *field.Path, allowSu
...
+			if isAcceptedAdmissionReviewVersion(v) {
+				hasAcceptedVersion = true
+			}
+			if seen[v] {
+				allErrors = append(allErrors, field.Invalid(fldPath.Index(i), v, "duplicate version"))
+			}
+			seen[v] = true
+		}
+		if requireRecognizedVersion && !hasAcceptedVersion {
+			allErrors = append(allErrors, field.Invalid(fldPath, versions, "none of the versions accepted by this server"))
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: nit: would be nice to include the versions this server does accept in the message

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 16]
```diff
@@ -225,8 +230,53 @@ func validateEnumStrings(fldPath *field.Path, value string, accepted []string, r
...
+}
+
+func validateConversionReviewVersions(versions []string, requireRecognizedVersion bool, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(versions) < 1 {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else {
+		seen := map[string]bool{}
+		hasAcceptedVersion := false
+		for i, v := range versions {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: limit version string to `utilvalidation.IsDNS1035Label`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 17]
```diff
@@ -225,8 +230,53 @@ func validateEnumStrings(fldPath *field.Path, value string, accepted []string, r
...
+			if isAcceptedConversionReviewVersion(v) {
+				hasAcceptedVersion = true
+			}
+			if seen[v] {
+				allErrs = append(allErrs, field.Invalid(fldPath.Index(i), v, "duplicate version"))
+			}
+			seen[v] = true
+		}
+		if requireRecognizedVersion && !hasAcceptedVersion {
+			allErrs = append(allErrs, field.Invalid(fldPath, versions, "none of the versions accepted by this server"))
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: include the accepted versions in the error message

**Comment by spiffxp** [comment]
/hold
v1.14 release lead here, this is a size/XXL PR that is attempting to land right before Code Freeze... are you _really sure_ this is a good idea?

**Comment by mbohlool** [comment]
@spiffxp I want to get this in. There are lots of generated and mechanical changes in this PR. The review is almost done.

**Comment by mbohlool** [comment]
@liggitt comment addressed.

**Comment by liggitt** [comment]
squashed, retagging
/lgtm
/hold cancel

**Comment by liggitt** [comment]
> v1.14 release lead here, this is a size/XXL PR that is attempting to land right before Code Freeze... are you _really sure_ this is a good idea?

The vast majority of the size is generated changes and thorough unit tests.

**Comment by liggitt** [comment]
flake https://github.com/kubernetes/kubernetes/issues/75159

/retest

**Comment by liggitt** [comment]
included in docs PR for 1.15 at https://github.com/kubernetes/website/pull/14671
